From 8534a1ba236ece814b99f58b45788404a2a2cd5e Mon Sep 17 00:00:00 2001
Message-Id: <8534a1ba236ece814b99f58b45788404a2a2cd5e.1427265330.git.chang-joon.lee@intel.com>
In-Reply-To: <b4c66aa6a04d2be0fca9fa4448c7188a581a57a5.1427265330.git.chang-joon.lee@intel.com>
References: <b4c66aa6a04d2be0fca9fa4448c7188a581a57a5.1427265330.git.chang-joon.lee@intel.com>
From: Deepak S <deepak.s@intel.com>
Date: Mon, 16 Mar 2015 15:44:41 +0530
Subject: [PATCH 14/22] SQUASHME! [VPG]: drm/i915: Optimize the runtime
 Promotion timer updates.

Avoid writing to the promotion timer if it is already programmed to the desired state.

SQUASHME! - This patch should be squashed into the following existing patch:
commit 680ecbaf6db3f2e5832f413198a8dd1f1e77c2fd
Change-Id: I18069d753f6e2a4735f86c0d203bbb971c9f62f7
Author: Deepak S <deepak.s@intel.com>
Date:   Thu Feb 20 21:42:37 2014 +0530

    FOR_UPSTREAM [VPG]: drm/i915: Modifying RC6 Promotion timer for
    Media workloads on VLV.

Issue: GMINL-7186
Change-Id: I0b0065604ad63275fb2bf4a857045785bc6f5023
Signed-off-by: Deepak S <deepak.s@intel.com>
---
 drivers/gpu/drm/i915/i915_drv.h |    3 +++
 drivers/gpu/drm/i915/intel_pm.c |    5 +++++
 2 files changed, 8 insertions(+)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 3098395..9dd8d07 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -1747,6 +1747,9 @@ struct drm_i915_private {
 	bool is_resuming;
 	bool is_video_playing;  /* Indicates enabling only in videomode */
 
+	/* Track the media promotion timer update */
+	bool last_media_active_state;
+
 	/* Indicates currently enabled planes */
 	unsigned int plane_stat;
 
diff --git a/drivers/gpu/drm/i915/intel_pm.c b/drivers/gpu/drm/i915/intel_pm.c
index a4cfedf..bf80f24 100644
--- a/drivers/gpu/drm/i915/intel_pm.c
+++ b/drivers/gpu/drm/i915/intel_pm.c
@@ -3931,6 +3931,9 @@ static void valleyview_disable_rps(struct drm_device *dev)
 void vlv_modify_rc6_promotion_timer(struct drm_i915_private *dev_priv,
 				    bool media_active)
 {
+	if (dev_priv->last_media_active_state == media_active)
+		return;
+
 	intel_runtime_pm_get(dev_priv);
 
 	/* Update RC6 promotion timers */
@@ -3947,6 +3950,8 @@ void vlv_modify_rc6_promotion_timer(struct drm_i915_private *dev_priv,
 	}
 
 	intel_runtime_pm_put(dev_priv);
+
+	dev_priv->last_media_active_state = media_active;
 }
 
 static void vlv_media_timeout_work_func(struct work_struct *work)
-- 
1.7.9.5

